nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1188,28 +1167,102 @@ impl DataStore {
                 );
             }
         }
-        //
-        // NOTE: from this point forward it is OK if we update the
-        // instance's `boot_disk_id` column with the updated value
-        // again. It will have already been assigned with constraint
-        // checking performed above, so updates will just be
-        // repetitive, not harmful.
 
-        let r = diesel::update(instance_dsl::instance)
+        let query = diesel::update(instance_dsl::instance)
+            .into_boxed()
             .filter(instance_dsl::id.eq(authz_instance.id()))
-            .filter(instance_dsl::state.eq_any(OK_TO_SET_BOOT_DISK_STATES))
+            .filter(instance_dsl::state.eq_any(InstanceState::NO_VMM_STATES));
+        let query = if boot_disk_id.is_some() {
+            query.filter(
+                instance_dsl::boot_disk_id
+                    .ne(boot_disk_id)
+                    .or(instance_dsl::boot_disk_id.is_null()),
+            )
+        } else {
+            query.filter(instance_dsl::boot_disk_id.is_not_null())
+        };
+
+        let r = query
             .set(instance_dsl::boot_disk_id.eq(boot_disk_id))
             .check_if_exists::<Instance>(authz_instance.id())
             .execute_and_check(&conn)
             .await?;
         match r.status {
             UpdateStatus::NotUpdatedButExists => {
-                // This should be the only reason the query would fail...
-                debug_assert!(!OK_TO_SET_BOOT_DISK_STATES
-                    .contains(&r.found.runtime().nexus_state));
-                Err(err.bail(Error::conflict(
-                    "instance must be stopped to set boot disk",
-                )))
+                if r.found.boot_disk_id == boot_disk_id {
+                    // Not updated, because the update is no change..
+                    return Ok(());
+                }
+
+                if !InstanceState::NO_VMM_STATES
+                    .contains(&r.found.runtime().nexus_state)
+                {
+                    return Err(err.bail(Error::conflict(
+                        "instance must be stopped to set boot disk",
+                    )));
+                }
```

Comment:
subtle but i think mildly neat detail: i moved this from being the fall-through error with a `debug_assert` to an explicit check because the bug in `instance_set_size_on_conn` was similar and would return a `conflict("instance must be stopped to be resized")` ... when the instance _was_ stopped and the failure to update was actually that i'd filtered on the cpu/memory condition incorrectly.

with this change we only tell users the instance must be stopped when we know affirmatively that it is not stopped, and the fallthrough case is truly just a way to say "this is a bug, on us somehow".